rmw_zenoh_cpp/src/detail/rmw_client_data.cpp

```diff
@@ -545,13 +542,41 @@ rmw_ret_t ClientData::shutdown()
   }
 
   is_shutdown_ = true;
+}
+
+///=============================================================================
+rmw_ret_t ClientData::shutdown()
+{
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
+  _shutdown();
   return RMW_RET_OK;
 }
 
+///=============================================================================
+bool ClientData::shutdown_and_query_in_flight()
+{
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
+  _shutdown();
+  return num_in_flight_ > 0;
+}
+
+///=============================================================================
+void ClientData::decrement_in_flight_and_conditionally_remove()
+{
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
+  --num_in_flight_;
+
+  if (is_shutdown_ && num_in_flight_ == 0) {
+    rmw_context_impl_s * context_impl = static_cast<rmw_context_impl_s *>(rmw_node_->data);
+    std::shared_ptr<rmw_zenoh_cpp::NodeData> node_data = context_impl->get_node_data(rmw_node_);
+    node_data->delete_client_data(rmw_client_);
```

Comment:
Could there be a possibility that the `context_impl` and `node_data` are nullptr? Eg. if for some reason the node is cleaned before we get the query reply. So we should check that these are not nullptr before proceeding with `get_node_data` and `delete_client_data`?
```suggestion
    rmw_context_impl_s * context_impl = static_cast<rmw_context_impl_s *>(rmw_node_->data);
    if (context_impl == nullptr) {
      return;
    }
    std::shared_ptr<rmw_zenoh_cpp::NodeData> node_data = context_impl->get_node_data(rmw_node_);
    if (node_data == nullptr) {
      return;
    }
    node_data->delete_client_data(rmw_client_);
```

Reply:
Yeah, really good call.  Fixed in [ec6c181](https://github.com/ros2/rmw_zenoh/pull/310/commits/ec6c181797d9d234f9577de6154ede4067d4b760)